token-analytics/internal/quicknode/quicknode.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/pkg/errors"
+)
+
+func NewClient(ctx context.Context, applicationYamlKey string) Client {
+	var cfg config
+	appcfg.MustLoadFromKey(applicationYamlKey, &cfg)
+	if cfg.QuickNode.APIKey == "" || cfg.QuickNode.APIKey == "-" {
+		cfg.QuickNode.APIKey = os.Getenv("QUICKNODE_API_KEY")
+	}
+	conf, err := pgxpool.ParseConfig(cfg.QuickNode.StreamDestinationURL)
+	if err != nil {
+		log.Panic(errors.Wrapf(err, "failed to parse destination url"))
+	}
+	q := &client{
+		bondingCurveSmartContractFilterTemplate: template.Must(template.New("bondingCurveSmartContractTemplate").Parse(bondingCurveSmartContractTemplate)),
+		httpClient:                              req.C().SetBaseURL("https://api.quicknode.com/"),
+		config:                                  &cfg,
+		streamDestination:                       conf.ConnConfig,
+	}
+	if err = q.bootstrap(ctx); err != nil {
+		log.Panic(errors.Wrapf(err, "failed to bootstrap quick node conn"))
+	}
+	return q
+}
+
+func (q *client) bootstrap(ctx context.Context) error {
+	if resp, err := q.req(ctx).Get("/streams/rest/v1/streams"); err != nil {
+		return errors.Wrapf(err, "failed to get /streams/rest/v1/streams")
+	} else if resp.GetStatusCode() >= http.StatusBadRequest {
+		return errors.Errorf("failed get /streams/rest/v1/streams with %v", resp.GetStatusCode())
+	} else if _, err2 := resp.ToBytes(); err2 != nil {
+		return errors.Wrapf(err2, "failed to read body of /streams/rest/v1/streams")
+	} else {
+		return nil
+	}
+}
+
+func (q *client) req(ctx context.Context) *req.Request {
+	return q.httpClient.R().
+		SetContext(ctx).
+		SetRetryBackoffInterval(100*time.Millisecond, 10*time.Second).
+		SetRetryHook(func(resp *req.Response, err error) {
+			switch {
+			case err != nil:
+				log.Error(errors.Wrapf(err, "faied to exec quick node request %v", resp.Request.URL.String()))
+			case resp.GetStatusCode() >= http.StatusBadRequest:
+				log.Error(errors.Errorf("quick node request failed %v: %v", resp.Request.URL.String(), resp.GetStatusCode()))
+			}
+		}).
+		SetRetryCount(5).
+		SetRetryCondition(func(resp *req.Response, err error) bool {
+			return err != nil || resp.GetStatusCode() >= http.StatusBadRequest
+		}).
+		SetHeader("x-api-key", q.config.QuickNode.APIKey)
+}
+
+func (q *client) CreateStream(ctx context.Context, streamName, contractAddrToMonitor string) (*Stream, error) {
+	filter, err := q.bondingCurveSmartContractFilterFunc(contractAddrToMonitor)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load filter function")
+	}
+	sslmode := "disable"
+	if strings.Contains(q.config.QuickNode.StreamDestinationURL, "sslmode=require") {
+		sslmode = "require"
+	}
+	params := &createStreamReq{
+		Name:                  streamName,
+		Network:               q.config.QuickNode.Network,
+		Dataset:               "block_with_receipts",
+		FilterFunction:        base64.StdEncoding.EncodeToString(filter),
+		Region:                "usa_east",
```

Comment:
has to be as close to the master postgres as possible; so probably we need to set it as an cfg value?